l10n_fr_department/model/res_partner.py

```diff
@@ -18,33 +20,35 @@ class ResPartner(models.Model):
     @api.depends("zip", "country_id", "country_id.code")
     # If a department code changes, it will have to be manually recomputed
     def _compute_department(self):
-        rcdo = self.env["res.country.department"]
+        def _get_zipcode(p) -> str:
+            if not p.country_id.id or p.country_id.id not in fr_country_ids:
+                return ""
+            p_zip = p.zip
+            if not p_zip or len(p_zip) != 5:
+                return ""
+            return p_zip.strip().replace(" ", "").rjust(5, "0")
+
         fr_country_ids = (
             self.env["res.country"]
             .search([("code", "in", ("FR", "GP", "MQ", "GF", "RE", "YT"))])
             .ids
         )
-        for partner in self:
+        for zipcode, partner_list in groupby(self, key=_get_zipcode):
             dpt_id = False
-            zipcode = partner.zip
-            if (
-                partner.country_id
-                and partner.country_id.id in fr_country_ids
-                and zipcode
-                and len(zipcode) == 5
-            ):
-                zipcode = partner.zip.strip().replace(" ", "").rjust(5, "0")
+            if zipcode:
                 code = self._fr_zipcode_to_department_code(zipcode)
-                dpt = rcdo.search(
-                    [
-                        ("code", "=", code),
-                        ("country_id", "in", fr_country_ids),
-                    ],
-                    limit=1,
+                dpt_id = (
+                    self.env["res.country.department"]
+                    .search(
+                        [("code", "=", code), ("country_id", "in", fr_country_ids)],
+                        limit=1,
+                    )
+                    .id
                 )
-                dpt_id = dpt and dpt.id or False
-            partner.department_id = dpt_id
+            self.browse().concat(*partner_list).department_id = dpt_id
```

Comment:
If you wanna be even more efficient, you could move the search out of the loop.
Something like this:

```py
        partners_by_zip = groupby(self, key=_get_zipcode)
        department_codes = [
            self._fr_zipcode_to_department_code(zipcode)
            for zipcode in partners_by_zip.keys()
            if zipcode
        ]
        departments = (
            self.env["res.country.department"].search(
                [("code", "in", department_codes), ("country_id", "in", fr_country_ids)]
            )
            if department_codes
            else self.env["res.country.department"]
        )
        departments_by_code = groupby(departments, key=lambda d: d.code)
        for zipcode, partner_list in groupby(self, key=_get_zipcode):
            department = self.env["res.country.department"]
            if zipcode:
                code = self._fr_zipcode_to_department_code(zipcode)
                department = departments_by_code[code][0]
            self.browse().concat(*partner_list).department_id = department
```

Note that `_fr_zipcode_to_department_code` is called twice per zipcode, however thanks to the `ormcache` the second time will be super quick.

However, we could group the partners by department code instead of grouping by zip, and so we wouldn't need to call this method twice